src/dune_rules/action_unexpanded.ml

```diff
@@ -367,15 +379,76 @@ end = struct
       register_dep b env acc ~f:(function
         | Ok p, _ -> Some p
         | Error _, _ -> None)
+
+    let ppx ~context ~expander ~scope ~loc ~lib_name ~ml_kind ~src pps flags env
+        acc =
+      let open Memo.O in
+      let* src, acc = dep src env acc in
+      let b =
+        let open Action_builder.O in
+        let corrected_suffix = ".ppx-corrected" in
+        let+ exe, ppx_flags, args =
+          Action_builder.memoize
+            ~cutoff:
+              (Tuple.T3.equal Path.Build.equal (List.equal String.equal)
+                 (List.equal String.equal))
+            "ppx driver and flags"
+            (let* () = Action_builder.return () in
+             let* exe, driver, flags =
+               let pps =
+                 List.map pps ~f:(fun (loc, lib) ->
+                     (loc, Lib_name.parse_string_exn (loc, lib)))
+               in
+               Ppx_driver.ppx_exe_driver_and_flags ~context ~expander ~loc
+                 ~lib_name ~flags ~scope pps
+             in
+             let+ ppx_flags =
+               let info = Ppx_driver.Driver.info driver in
+               Ppx_driver.driver_flags expander ~corrected_suffix
+                 ~driver_flags:info.flags
+                 ~standard:(Action_builder.return [ "--as-ppx" ])
+             in
+             (exe, ppx_flags, flags))
+        and+ src =
+          src >>= function
+          | External p -> Action_builder.return (Path.External.to_string p)
+          | In_source_tree _ -> assert false
+          | In_build_dir b ->
+            (* PPX runs at the workspace root *)
+            let+ dir =
+              String_with_vars.virt_pform __POS__ (Var Workspace_root)
+              |> Expander.expand_path env
```

Comment:
I think this is just `context.build_dir`.

Reply:
Actually, is this the right place to do this? It's the `preprocess` field that is responsible for setting `chdir` for normal preprocessors. I don't see why this is any different.

Reply:
I just need to find the path relative to where the PPX is running, to find the source file. What's the best way to do that? I can use` context.build_dir`

Reply:
I guess there's no way to do that here because there could be a `(chdir ..)` wrapping the action later. Perhaps we should just hard code the chdir action here and expand relative to context.build_dir?

In the future, we really need to stop flattening paths inside actions that aren't wrapped with a `(chdir ..)`

Reply:
This is a general problem we have in many other cases, so there's no need for a principled fix here.